java/spark-1.5.2/rutils.scala

```diff
@@ -0,0 +1,161 @@
+package sparklyr
+
+import java.io.ByteArrayInputStream
+import java.io.DataInputStream
+import java.nio.charset.StandardCharsets
+
+object RUtils {
+  private[this] val XDR_FORMAT = 'X'
+  private[this] val CHARSXP = 9
+  private[this] val LGLSXP = 10
+  private[this] val INTSXP = 13
+  private[this] val REALSXP = 14
+  private[this] val STRSXP = 16
+  private[this] val VECSXP = 19
+  private[this] val NILVALUE_SXP = 254
+  private[this] val NA_INTEGER = Integer.MIN_VALUE
+
+  def validateSerializationFormat(
+    dis: DataInputStream,
+    verbose: Boolean = false
+  ): Unit = {
+    val format = Serializer.readObjectType(dis)
+    dis.readByte()  // skip the 2nd byte which we currently do not use
+    if (XDR_FORMAT != format)
+      throw new IllegalArgumentException(
+        s"Unsupported R serialization format '$format'"
+      )
+
+    val version = Serializer.readInt(dis)
+    val writerVersion = Serializer.readInt(dis)
+    val minReaderVersion = Serializer.readInt(dis)
+
+    if (version != 2)
+      throw new IllegalArgumentException(
+        s"Unsupported R serialization version '$version'"
+      )
+
+    if (verbose) {
+      val logger = new Logger("Utils", 0)
+      logger.log(s"R serialization version: $version")
+      logger.log(s"R serialization writer version: $writerVersion")
+      logger.log(s"R serialization min reader version: $minReaderVersion")
+    }
+  }
+
+  def unserializeColumn(dis: DataInputStream) : Array[Any] = {
+    val logger = new Logger("RUtils", 0)
+
+    val dtype = readDataType(dis) // read column type
+    if (VECSXP != dtype)
+      logger.logWarning(s"Unexpected column data type $dtype")
+
+    val num_rows = readLength(dis)
+
+    (0L until num_rows).map(
+      r => {
+        val elem_dtype = readDataType(dis)
+
+        if (NILVALUE_SXP == elem_dtype) {
+          null
+        } else {
+          val num_elems = readLength(dis)
+
+          if (1 != num_elems)
+            throw new IllegalArgumentException(
+              s"Unexpected number of elements: $num_elems"
+            )
+
+          elem_dtype match {
+            case LGLSXP => {
+              val v = Serializer.readInt(dis)
+              if (NA_INTEGER == v) null else v != 0
+            }
+            case INTSXP => {
+              val v = Serializer.readInt(dis)
+              if (NA_INTEGER == v) null else v
+            }
+            case REALSXP => {
+              val bytes = new Array[Byte](8)
+              dis.readFully(bytes)
+              val hw: Long = extractInt(bytes, 0) & 0x00000000FFFFFFFFL
+              val lw: Long = extractInt(bytes, 4) & 0x00000000FFFFFFFFL
```

Comment:
You'll see a number of these bitwise `&` operators here because there is no unsigned integer types in Java/Scala and I need to mask high-order bits that could become all `1`s after sign extension.